main.py

```diff
@@ -1,30 +1,91 @@
 """
 This is an Add-On for DocumentCloud which wraps the tabula-py library
 https://github.com/chezou/tabula-py
-
-It attempts to convert tables from PDFs to CSVs
+It allows you to export tables from a set of PDFs into CSVs and
+download the resulting CSVs as a zip file.
 """
 
+import os
+import sys
+import fnmatch
+import zipfile
+import subprocess
+from urllib.parse import urlparse
+import requests
 import tabula
 from documentcloud.addon import AddOn
-
+from clouddl import grab
 
 class Tabula(AddOn):
-    """A tabula Add-On for DocumentCloud"""
-
-    def main(self):
-
-        document = self.client.documents.get(self.documents[0])
-        with open(f"{document.slug}.pdf", "wb") as pdf_file:
-            pdf_file.write(document.pdf)
+    """A tabula PDF table extraction Add-On for DocumentCloud"""
 
-        tabula.convert_into(
-            f"{document.slug}.pdf", f"{document.slug}.csv", output_format="csv", pages="all"
-        )
+    def fetch_template(self, url):
+        """Fetch the template from either Dropbox, Google Drive, or any public URL"""
+        os.makedirs(os.path.dirname("./out/"), exist_ok=True)
+        if(grab(url, "./out/")):
+            for file in os.listdir('./out/'):
+                if fnmatch.fnmatch(file, '*.json'):
+                   template_path = os.path.join('./out/', file)
+                   os.rename(template_path, os.path.join('./out/', 'template.json'))
+                   return True
+        else:
+            parsed_url = urlparse(url)
+            basename = os.path.basename(parsed_url.path)
+            title, ext = os.path.splitext(basename)
+            if not title:
+                title = "template"
+            if not ext:
+                ext = "json"
+            with requests.get(url, stream=True, timeout=10) as resp:
+                if resp.status_code == 200:
+                    with open("./out/template.json", "wb") as json_file:
+                        for chunk in resp.iter_content(chunk_size=8192):
+                            json_file.write(chunk)
+                    return True
+                else:
+                    return False
+           
+    def template_based_extract(self, url):
+        if(self.fetch_template(url)):
+            with zipfile.ZipFile("export.zip", mode="w") as archive:
+                for document in self.get_documents():
+                    with open("file.pdf", "wb") as pdf_file:
+                        pdf_file.write(document.pdf)
+                    data_frame_list = tabula.read_pdf_with_template("./file.pdf", "./out/template.json") 
+                    # Tabula's read_pdf_with_template() returns a list of data frames we can append to form a CSV. 
+                    for data_frame in data_frame_list:
+                        data_frame.to_csv(f"{document.slug}.csv", mode="a", index=False, header=False)
+                    archive.write(f"{document.slug}.csv")
+        else:
```

Comment:
Again, I think there should be some message here, and possibly it failing to run rather than falling back to not using the template, if they are expecting a template to be used.